src/gtil_utils.cu

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <algorithm>
+#include <cstring>
+
+#include <treelite/c_api.h>
+#include <treelite/tree.h>
+#include <triton_fil/gtil_utils.cuh>
+#include <triton_fil/model_instance_state.cuh>
+#include <triton_fil/model_state.h>
+#include <triton_fil/triton_tensor.cuh>
+
+namespace triton { namespace backend { namespace fil {
+void
+gtil_predict(
+    ModelInstanceState& instance_state, TritonTensor<const float>& data,
+    TritonTensor<float>& preds, bool predict_proba)
+{
+  auto& model_state = [&]() -> ModelState& {
+    auto model_ptr = instance_state.StateForModel();
+    if (model_ptr == nullptr) {
+      throw TritonException(
+          TRITONSERVER_errorcode_enum::TRITONSERVER_ERROR_INTERNAL,
+          "No model associated with instance");
+    }
+    return *model_ptr;
+  }();
+
+  std::size_t num_class = 1;
+  auto tl_model = static_cast<treelite::Model*>(model_state.treelite_handle);
+  num_class = tl_model->task_param.num_class;
+  if (!predict_proba && model_state.tl_params.output_class && num_class > 1) {
+    std::strcpy(tl_model->param.pred_transform, "max_index");
+  }
+
+  std::size_t out_result_size;
+  int gtil_result = TreeliteGTILPredict(model_state.treelite_handle, data.data(),
+              data.shape()[0], preds.data(), 1, &out_result_size);
```

Comment:
For a multi-class classifier with `predict_proba=False`, GTIL requires the buffer of size `num_row * num_class` to store the raw scores. So in this case `out_result_size == num_row < num_row * num_class`